misk-feature-testing/src/main/kotlin/misk/feature/testing/FakeFeatureFlags.kt

```diff
@@ -77,76 +77,149 @@ class FakeFeatureFlags @Inject constructor(private val moshi : Provider<Moshi>)
   override fun <T : Enum<T>> getEnum(feature: Feature, clazz: Class<T>): T = getEnum(feature, KEY, clazz)
   override fun <T> getJson(feature: Feature, clazz: Class<T>): T = getJson(feature, KEY, clazz)
 
-  private fun get(feature: Feature, key: String): Any? {
+  private fun <V> getOrDefault(
+    feature: Feature,
+    key: String,
+    defaultValue: V,
+    attributes: Attributes = defaultAttributes
+  ): V {
     FeatureFlagValidation.checkValidKey(feature, key)
-    return overrides.getOrElse(MapKey(feature, key)) {
-      overrides[MapKey(feature)]
-    }
+    @Suppress("unchecked_cast")
+    return (get(feature, key, attributes) ?: defaultValue) as V
   }
 
-  private fun <V> getOrDefault(feature: Feature, key: String, defaultValue: V): V {
+  private fun get(
+    feature: Feature,
+    key: String,
+    attributes: Attributes = defaultAttributes
+  ): Any? {
     FeatureFlagValidation.checkValidKey(feature, key)
-    @Suppress("unchecked_cast")
-    return overrides.getOrElse(MapKey(feature, key)) {
-      overrides.getOrDefault(MapKey(feature), defaultValue)
-    } as V
+    // Override value lookup is structured in 3 levels: feature, key, and attributes level. The
+    // logic to get the override value is as follows:
+    //
+    // 1. Return feature level override if there is no override at the key level.
+    // 2. if there is an override at the key level, then:
+    //   2.1 Return the override value associated with the attribute override that contains
+    //       all the attributes of the given attributes.
+    //   2.2 If there is no match, return the override value defined at the key level.
+    val overrideMapValues = overrides[MapKey(feature, key)]
+        ?: return overrides[MapKey(feature)]?.let { it.first().value }
+
+    val overrideMapValuesCopy = PriorityQueue(overrideMapValues)
```

Comment:
is this intending to defend against races? Or just so you can mutate?

Reply:
the latter, so I can do the `poll`. I didn't want to mutate the original data structure so that's why I am making a copy of it. The other option that I see is to mutate the original data structure, put the elements in a stack, and then when I am done put them back. (too much work without any much benefits).

Thoughts?